### PR TITLE
Review by HarveySouth

### PR DESCRIPTION
**Pull request by HarveySouth**
Place for review/discussion following from #15

### CODE REVIEW

**Comment by HarveySouth** [comment]
OK, I have managed to read through the notebooks and run through the tutorial start to finish without fail. I committed the changes I have made to make this possible, and will now annotate the changes in review (what and why)

File: 02_data_prep.ipynb [hunk 1]
```diff
@@ -128,7 +128,7 @@
   },
```
Comment (HarveySouth) [inline 02_data_prep.ipynb:133]: <p>The data from zenodo did not fit the expectation of the notebooks, I found that the name of the column was different: Rotors 1 is true vs expected rotors_present, and the values in the rows were different: the large majority of "no rotors" category was NaN instead of 0. Originally, I had updated the target_feature_name variable in each of the notebooks and ray_hpt, but the NaN caused training to work, but not as intended as the 'no rotors' category became empty. To solve this, I added this cell which changes the zenodo data and saves it back if the target feature name doesn't exist.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/ukrse_2022_mlops_walkthrough/pull/24/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='24'/>

Comment (HarveySouth) [inline 02_data_prep.ipynb:133]: <p>this change relates to an earlier comment made in #15 . the download link in README.md for the era5 data downloaded the rotors data instead of the era5 nc file, so I updated the readme and this line, it is possible to revert both changes so that the data directory does contain the era5 subdirectory, but this is not how it is currently implemented here. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/ukrse_2022_mlops_walkthrough/pull/24/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='65'/>

File: 03_model_development.ipynb [hunk 2]
```diff
@@ -118,7 +118,7 @@
   },
```
Comment (HarveySouth) [inline 03_model_development.ipynb:116]: <p>This was not a necessary addition, but will make a review-esque comment instead: </p><p> </p><p>I'm not super familiar with bash variables, but didn't have <span class="MathJax_Preview" style="color: inherit;">PORT or </span>ARTIFACT_PATH vars set to run the mlflow server command example, I included this example which works for me, and I thought was general enough to be useful to include (I am also not familiar with port usage, so can't think of a reason to change the port set at the top of the notebook) (the two data storage definitions (db + artifacts) are relative paths so the mlflow server is run within the repo directory). The notes above were helpful, but I think it would be useful to mention that the port definition in the notebook is found assigned at the top of the notebook, or guidance on setting the PORT or ARTIFACT_PATH variables (if possible) or even setting the PORT or ARTIFACT_PATH variables from within the notebook (if it is possible to make them accessible across terminal instances).</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/ukrse_2022_mlops_walkthrough/pull/24/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='43'/>

Comment (HarveySouth) [inline 03_model_development.ipynb:116]: <p>Changed the tensorboard logdir to be relative</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/ukrse_2022_mlops_walkthrough/pull/24/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='53'/>

Comment (HarveySouth) [inline 03_model_development.ipynb:116]: <p>changed the time magic to time the cell instead of the empty line</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/ukrse_2022_mlops_walkthrough/pull/24/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='54'/>

File: 04_model_evaluation.ipynb [hunk 3]
```diff
@@ -88,7 +88,7 @@
   },
```
Comment (HarveySouth) [inline 04_model_evaluation.ipynb]: <p>I am not too sure of a programmatic way to specify a logged model and assume that the previous variable set for logged model was hard coded for a run found mlflow. I changed it to show the latest run instead</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/ukrse_2022_mlops_walkthrough/pull/24/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='39'/>

Comment (HarveySouth) [inline 04_model_evaluation.ipynb]: <p>Another review-esque comment for my unnecessary addition:</p><p> </p><p>It took me some time to realise that the models have to be registered by the user of MLflow for them to return on this function call. I assume there might be some programmatic way to do this but ended up doing it through the mlflow UI as shown on the mlflow introduction documentation to registry. I think it would be useful to make clear in the above markdown that the model must be registered manually, or to register the model trained in the previous notebook programmatically after training, since currently only one model is logged to mlflow.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/ukrse_2022_mlops_walkthrough/pull/24/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='46'/>

Comment (HarveySouth) [inline 04_model_evaluation.ipynb]: <p>Since I registered the model manually, I changed this to the name I gave the model. Could be useful to update this with something programmatic or have this assignment at the top of the notebook as a variable setting.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/ukrse_2022_mlops_walkthrough/pull/24/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='58'/>

Comment (HarveySouth) [inline 04_model_evaluation.ipynb]: <p>For this cell and above, in cases where a conf mat was generated for a batch that only had one label, the dimensions of the output from the sklearn.metrics.confusion_matrix function became [1,1] instead of [2,2]. Passing the label argument to the function ensures shape. I assume this won't be an encountered issue, but had an exception while running through while the data for rotors had NaN values (commented in 02)</p><p> </p><p>numpy.unique(y_val).astype(int) is the first implemented way of generating the labels, so could do with some further consideration in case of edge case failure </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/ukrse_2022_mlops_walkthrough/pull/24/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='70'/>

File: requirements_model_development.yml [hunk 4]
```diff
@@ -14,6 +14,8 @@ dependencies:
   - imbalanced-learn
   - mlflow=1.20
   - intake
+  - onnx
```
Comment (HarveySouth) [inline requirements_model_development.yml]: The model development notebook imported onnx and tf2onnx, so I included them in the requirements file

File: requirements_evaluation.yml [hunk 5]
```diff
@@ -12,12 +11,10 @@ dependencies:
   - scikit-learn=0.24
   - tensorflow=2.8
   - mlflow=1.20
-  - holoviz
-  - mlflow=1.20
+  - panel
   - intake
+  - pip
```
Comment (HarveySouth) [inline requirements_evaluation.yml]: conda gives a warning when pip is used, but not listed as a dependency. So I added pip to all requirements files with pip installs

File: requirements_evaluation.yml [hunk 6]
```diff
@@ -12,12 +11,10 @@ dependencies:
   - scikit-learn=0.24
   - tensorflow=2.8
   - mlflow=1.20
-  - holoviz
-  - mlflow=1.20
```
Comment (HarveySouth) [inline requirements_evaluation.yml]: duplicate dependency

File: requirements_evaluation.yml [hunk 7]
```diff
@@ -1,7 +1,6 @@
 name: ukrse2022_mlops_model_evaluation
 channels:
   - conda-forge
-  - pyviz
```
Comment (HarveySouth) [inline requirements_evaluation.yml]: I also made the change to remove the pyviz channel which I found still was not available. replaced with the import of conda forge's panel instead, I didn't detect any changes to the notebook outputs with this change

**Review by HarveySouth — COMMENTED** [review]
annotation on conda env files

**Review by HarveySouth — COMMENTED** [review]
The .gitignore is added, it may ignore more than desired, since I ran mlflow within the repo and stored the objects here too. I do think the pycache and ipynb_checkpoints removal is necessary though.

File: README.md [hunk 8]
```diff
@@ -16,34 +16,33 @@ From the command line, run the following command:
...
 The following environments should be installed for this walkthrough:
 * Data Preparation and Exploration - `requirements_data_prep.yml`
-* Model Development and Training - `requiremments_model_development.yml`
+* Model Development and Training - `requirements_model_development.yml`
 * Model Inference, Evaluation and Explainability (XAI) - `requirements_model_evaluation.yml`
 
 ### Get the data
 
-The data used in this walkthrough as an [archive on Zenodo](https://doi.org/10.5281/zenodo.6966936). Unformatuantely you can't easily just download all the files in a zendodo record, so you will need to download each. This can be done with the wget commands listed below. It is recommended that these be placed in a directory in ~/data/ukrse2022. 
+The data used in this walkthrough as an [archive on Zenodo](https://doi.org/10.5281/zenodo.6966936). Unfortunately you can't easily just download all the files in a Zenodo record, so you will need to download each. This can be done with the wget commands listed below. It is recommended that these be placed in a directory in ~/data/ukrse2022.
```
Comment (HarveySouth) [inline README.md]: Not sure if this wording was intentional, leaving a comment for review

File: README.md [hunk 9]
```diff
@@ -16,34 +16,33 @@ From the command line, run the following command:
...
 * Raw Rotors dataset `wget https://zenodo.org/record/6966937/files/2021_met_office_aviation_rotors.csv -P  ~/data/ukrse2022/`
-* Preprocessed rotors dataset `wget [XXX](https://zenodo.org/record/7022648/files/2021_met_office_aviation_rotors_preprocessed.csv -P ~/data/ukrse2022/`
-* Intake catalog file `wget [catalog.yml](https://raw.githubusercontent.com/informatics-lab/ukrse_2022_mlops_walkthrough/main/rotors_catalog.yml) -P  ~/data/ukrse2022/`
-* Preprocessed UK cutout of ERA5 Mean Sea-level Pressure dataset for 2017-2021 `wget https://zenodo.org/record/6966937/files/2021_met_office_aviation_rotors.csv -P ~/data/ukrse2022/`
-* 
+* Preprocessed rotors dataset `wget https://zenodo.org/record/7022648/files/2021_met_office_aviation_rotors_preprocessed.csv -P ~/data/ukrse2022/`
+* Intake catalog file `wget https://raw.githubusercontent.com/informatics-lab/ukrse_2022_mlops_walkthrough/main/rotors_catalog.yml -P  ~/data/ukrse2022/`
+* Preprocessed UK cutout of ERA5 Mean Sea-level Pressure dataset for 2017-2021 `wget https://zenodo.org/record/7022648/files/era5_mslp_UK_2017_2020.nc -P ~/data/ukrse2022/`
```
Comment (HarveySouth) [inline README.md]: Updated the ERA5 data to point to the era5 .nc instead of the rotors data. the path is ~/data/ukrse2022/ instead of the possibly intended /data/era5 . also related to the path definition found in notebook 02.

File: README.md [hunk 10]
```diff
@@ -16,34 +16,33 @@ From the command line, run the following command:
...
 * Raw Rotors dataset `wget https://zenodo.org/record/6966937/files/2021_met_office_aviation_rotors.csv -P  ~/data/ukrse2022/`
-* Preprocessed rotors dataset `wget [XXX](https://zenodo.org/record/7022648/files/2021_met_office_aviation_rotors_preprocessed.csv -P ~/data/ukrse2022/`
-* Intake catalog file `wget [catalog.yml](https://raw.githubusercontent.com/informatics-lab/ukrse_2022_mlops_walkthrough/main/rotors_catalog.yml) -P  ~/data/ukrse2022/`
-* Preprocessed UK cutout of ERA5 Mean Sea-level Pressure dataset for 2017-2021 `wget https://zenodo.org/record/6966937/files/2021_met_office_aviation_rotors.csv -P ~/data/ukrse2022/`
-* 
+* Preprocessed rotors dataset `wget https://zenodo.org/record/7022648/files/2021_met_office_aviation_rotors_preprocessed.csv -P ~/data/ukrse2022/`
+* Intake catalog file `wget https://raw.githubusercontent.com/informatics-lab/ukrse_2022_mlops_walkthrough/main/rotors_catalog.yml -P  ~/data/ukrse2022/`
```
Comment (HarveySouth) [inline README.md]: The two links here didn't run as wget commands, I removed the markdown-like link embedding syntax around the links in the commands

File: README.md [hunk 11]
```diff
@@ -16,34 +16,33 @@ From the command line, run the following command:
...
 * Data Preparation and Exploration - `requirements_data_prep.yml`
-* Model Development and Training - `requiremments_model_development.yml`
+* Model Development and Training - `requirements_model_development.yml`
 * Model Inference, Evaluation and Explainability (XAI) - `requirements_model_evaluation.yml`
 
 ### Get the data
 
-The data used in this walkthrough as an [archive on Zenodo](https://doi.org/10.5281/zenodo.6966936). Unformatuantely you can't easily just download all the files in a zendodo record, so you will need to download each. This can be done with the wget commands listed below. It is recommended that these be placed in a directory in ~/data/ukrse2022. 
+The data used in this walkthrough as an [archive on Zenodo](https://doi.org/10.5281/zenodo.6966936). Unfortunately you can't easily just download all the files in a Zenodo record, so you will need to download each. This can be done with the wget commands listed below. It is recommended that these be placed in a directory in ~/data/ukrse2022. 
 * Raw Rotors dataset `wget https://zenodo.org/record/6966937/files/2021_met_office_aviation_rotors.csv -P  ~/data/ukrse2022/`
```
Comment (HarveySouth) [inline README.md]: Something I didn't change, but the doi for this zenodo link is different from the others (6966937 vs 7022648)

**Review by HarveySouth — COMMENTED** [review]
Annotated README changes

**Comment by HarveySouth** [comment]
Next I will transfer the review comments from #15 to here for ease of discussion, then update notebooks on typos/wording and attempt to give some more review comments

**Comment by HarveySouth** [comment]
One possible solution to remove the empty holoviz panel cells that show during the initialization of a servable plot would be to use `from IPython.display import clear_output` and clear the cell output at the end of the cell with the empty plot

The same could be applied to the dask compute() `distributed.worker_memory - WARNING - Unmanaged memory use is high` spam

**Comment by HarveySouth** [comment]
Similar comments to https://github.com/informatics-lab/ml_weather_tutorial/pull/21#discussion_r990203038 (and surrounding) on overlapping text for plots, scale similarity and plt.show()

File: 02_data_prep.ipynb (hunk 1 above)
Comment (HarveySouth) [inline 02_data_prep.ipynb:133]: <p>The command line addition I will make here I believe is the correct command for the command line (link below). it doesn't seem to work, my assumption from the error message would be that dask doesn't interact well with the application, so when trying to load the task graph accidentally uses the string of an object instead of the correct object content itself for loading...</p><p> </p><p><a href="https://panel.holoviz.org/user_guide/Overview.html#panel-serve-app-ipynb" target="_blank">Overview — Panel v0.14.0 (holoviz.org)</a></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/ukrse_2022_mlops_walkthrough/pull/24/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='59'/>

**Comment by HarveySouth** [comment]
ray_results is saved to ~/ray_results where ~/data/ukrse2022/ray_results might be a more expected location

File: 03_model_development.ipynb (hunk 2 above)
Comment (stevehadd) [reply 03_model_development.ipynb:116]: <p>those variable we intends as placeholder, which is to say you can decide what port and path to use, rather than those are things which are already defined. I've created a script now to simplfy running mlflow in a single command so hopefully thisn will be clearer.</p>

File: 02_data_prep.ipynb (hunk 1 above)
Comment (stevehadd) [reply 02_data_prep.ipynb:133]: <p>thanks, I probably to re upload the data, the version got confusing. The notebooks all ran on spice now for me, so thanks for this addtition I'll look at a better fix after the cop session on monday</p>

Comment (stevehadd) [reply 02_data_prep.ipynb:133]: <p>thanks</p>

File: 04_model_evaluation.ipynb (hunk 3 above)
Comment (stevehadd) [reply 04_model_evaluation.ipynb]: <p>good point, i'll create an issue</p>

Comment (stevehadd) [reply 04_model_evaluation.ipynb]: <p>I think the good practice is mostly to be registering models you intend subsequent users to refer too. You can see an example of this further down. Otherwise for unregistered models a good pattern is searching for the best model by some metric and using that as also demonstrated below.</p>

Comment (stevehadd) [reply 04_model_evaluation.ipynb]: <p>yes, i realised in rerunning this. I'm going to use this to demo the GUI in the live tutorial, but will create an issue to also include a programmtic version #26</p>
